p2p/net/swarm/black_hole_detector.go

```diff
@@ -0,0 +1,208 @@
+		// If the call succeeds in a blocked state we reset to allowed.
+		// This is better than slowly accumulating values till we cross the minSuccessFraction
+		// threshold since a blackhole is a binary property.
+		b.reset()
+		return
+	}
+
+	if success {
+		b.successes++
+		b.outcomes = append(b.outcomes, outcomeSuccess)
+	} else {
+		b.failures++
+		b.outcomes = append(b.outcomes, outcomeFailed)
+	}
+
+	if len(b.outcomes) > b.n {
+		if b.outcomes[0] == outcomeSuccess {
+			b.successes--
+		} else {
+			b.failures--
+		}
+		b.outcomes = b.outcomes[1 : b.n+1]
+	}
+
+	b.updateState()
+	b.trackMetrics()
+}
+
+func (b *blackHoleFilter) IsAllowed() (state blackHoleState, isAllowed bool) {
+	if b == nil {
+		return blackHoleStateAllowed, true
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.requests++
+	if b.requests == b.n {
+		b.requests = 0
+	}
+	b.trackMetrics()
+	return b.state, (b.state == blackHoleStateAllowed) || (b.requests == 0)
+}
+
+func (b *blackHoleFilter) reset() {
+	b.successes = 0
+	b.failures = 0
+	b.outcomes = b.outcomes[:0]
+	b.updateState()
+}
+
+func (b *blackHoleFilter) updateState() {
+	st := b.state
+	successFraction := 0.0
+	if len(b.outcomes) < b.n {
+		b.state = blackHoleStateAllowed
+	} else {
+		successFraction = float64(b.successes) / float64(b.successes+b.failures)
+		if successFraction >= b.minSuccessFraction {
+			b.state = blackHoleStateAllowed
+		} else {
+			b.state = blackHoleStateBlocked
+		}
+	}
+	if st != b.state {
+		if b.state == blackHoleStateAllowed {
+			log.Debugf("%s blackHoleDetector state changed to Allowed", b.name)
+		} else {
+			log.Debugf("%s blackHoleDetector state changed to Blocked. Success fraction is %0.3f", b.name, successFraction)
+		}
+	}
+}
+
+func (b *blackHoleFilter) trackMetrics() {
+	if b.metricsTracer == nil {
+		return
+	}
+	successFraction := 0.0
+	if b.successes+b.failures != 0 {
+		successFraction = float64(b.successes) / float64(b.successes+b.failures)
+	}
+	b.metricsTracer.UpdatedBlackHoleFilterState(
+		b.name,
+		b.state,
+		b.n-b.requests,
+		successFraction,
+	)
+}
+
+// blackHoleDetector provides UDP and IPv6 black hole detection using a `blackHoleFilter`
+// for each. For details of the black hole detection logic see `blackHoleFilter`
+type blackHoleDetector struct {
+	udp, ipv6 *blackHoleFilter
+}
+
+func (d *blackHoleDetector) IsAllowed(addr ma.Multiaddr) bool {
+	if !manet.IsPublicAddr(addr) {
+		return true
+	}
+
+	udpState, udpAllowed := blackHoleStateAllowed, true
+	if d.udp != nil && isProtocolAddr(addr, ma.P_UDP) {
+		udpState, udpAllowed = d.udp.IsAllowed()
+	}
+
+	ipv6State, ipv6Allowed := blackHoleStateAllowed, true
+	if d.ipv6 != nil && isProtocolAddr(addr, ma.P_IP6) {
+		ipv6State, ipv6Allowed = d.ipv6.IsAllowed()
+	}
+
+	// Allow all probes irrespective of the state of the other filter
+	if (udpState == blackHoleStateBlocked && udpAllowed) ||
+		(ipv6State == blackHoleStateBlocked && ipv6Allowed) {
+		return true
+	}
```

Comment:
I'm using the term probe to mean we are blocked and we want to reevaluate the black hole state by allowing the request.
This is a composite check. We want to allow when either of the blocked black hole filters decide that it is time to probe.  
Say UDP and IPv6 are both blocked and UDP black hole filter allows the probe but IPv6 black hole filter blocks it. Just a single check `udpAllowed && ipv6Allowed` will block this probe. I think it is better to allow all probes in case the black hole filter state has changed.

Reply:
I am now returning a `blackHoleResult` type which will be `blackHoleResultProbe` in this case.